imgpkg/002-recursive-bundles/README.md

```diff
@@ -0,0 +1,304 @@
+
+```=
+imgpkg copy -b ghcr.io/k14s/design-docs/simple-app-install-package --to-repo some-other.registry.io/recursive-bundle
+copy | exporting 4 images...
+copy | will export ghcr.io/k14s/design-docs/simple-app@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0
+copy | will export ghcr.io/k14s/design-docs/utilities@sha256:47ae428a887c41ba0aedf87d560eb305a8aa522ffb80ac1c96a37b16df038e0f
+copy | will export ghcr.io/k14s/design-docs/simple-app-bundle@sha256:d211dd700949154e429d28661d01c99d53a38af0d5275842ccbf0bf6dbef8ca4
+copy | will export ghcr.io/k14s/design-docs/simple-app-install-package@sha256:77c97e82306fb2a616da0a78796db039aa76e5bac1508954d40c0c10c073e035
+copy | exported 4 images
+copy | importing 4 images...
+copy | importing ghcr.io/k14s/design-docs/simple-app@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0 -> some-other.registry.io/recursive-bundle@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0...
+copy | importing ghcr.io/k14s/design-docs/utilities@sha256:ghcr.io/k14s/design-docs/utilities@sha256:47ae428a887c41ba0aedf87d560eb305a8aa522ffb80ac1c96a37b16df038e0f -> some-other.registry.io/recursive-bundle@sha256:ghcr.io/k14s/design-docs/utilities@sha256:47ae428a887c41ba0aedf87d560eb305a8aa522ffb80ac1c96a37b16df038e0f...
+copy | importing ghcr.io/k14s/design-docs/simple-app-bundle@sha256:d211dd700949154e429d28661d01c99d53a38af0d5275842ccbf0bf6dbef8ca4 -> some-other.registry.io/recursive-bundle@sha256:d211dd700949154e429d28661d01c99d53a38af0d5275842ccbf0bf6dbef8ca4...
+copy | importing ghcr.io/k14s/design-docs/simple-app-install-package@sha256:77c97e82306fb2a616da0a78796db039aa76e5bac1508954d40c0c10c073e035 -> some-other.registry.io/recursive-bundle@sha256:77c97e82306fb2a616da0a78796db039aa76e5bac1508954d40c0c10c073e035...
+copy | imported 4 images
+Succeeded
+```
+
+## Pull a bundle
+
+### Required changes
+
+#### Change folder structure
+
+When downloading a bundle that contains other bundles to disk using the `pull` command will work as currently, except for that it will download the nested bundles into a hidden folder called `.bundles`.
+
+##### Folder structure
+
+The structure of the output folder will be
+```
+$ tree -a recursive-bundle
+.
+├── .bundles
+│   ├── sha256-{SHA Of the First Nested Bundle}
+│   │   ├── .imgpkg
+│   │   │   ├── bundle.yml
+│   │   │   └── images.yml
+│   │   └── config2.yml
+│   └── sha256-{SHA Of the Second Nested Bundle}
+│   │   ├── .imgpkg
+│   │   │   ├── bundle.yml
+│   │   │   └── images.yml
+│       └── config1.yml
+├── .imgpkg
+│   ├── bundle.yml
+│   └── images.yml
+└── config.yml
+
+7 directories, 9 files
+```
+
+The folder name will be the sha256-{SHA} where SHA is the SHA256 of the bundle OCI Image. The mapping between the folder names and the origin images can be found in the `.imgpkg/images.yml` of the bundle that included this bundle
```

Comment:
Why dont we just put the image name + sha (maybe even just image@sha256... from the lockfile) as the filename instead of forcing users to manually decode which digest is which application bundle? The digest in the filename will ensure there are no collisions, and, if there are, we deduped our pull wrong.

Reply:
I have some concerns about the image name + sha approach.
- What do we consider image name? Only the repository or the full image name with the registry?
  - If we use only the repository name we might have to do some decoding by hand because we might have multiple images in the same repository
  - If we have the full image with registry this will become a pretty long name and we would also have problems with the sub-folder structure that we would have to convert the / -> _ or something of the sorts


The intention behind the `imgpkg info -b` is to also help with this particular problem where you could easily look at the image name and its sha so you could do the translation.

Reply:
Yea, thats true. I guess since the users will no longer have the context of where a certain bundle was referenced from, it is a bit harder to distinguish what the name means, i.e. knowing the parent bundle could helps distinguish overlapping names.

It just feels like a poor experience to have to find a specific digest in a list of potentially hundreds of images in order to find the app I'm looking for on disk.

Also, will the info command recursively compile a list of ALL images/bundles referenced by a bundle, and, if so, do we have an idea for the scale we expect? Will it also be able to provide me info from an on disk bundle? If the scale is huge and I just spent a significant amount of time pulling everything, I probably wouldn't want to wait for imgpkg to essentially pull everything again just to create the list of images for the info command, especially if I need to do that every time I'm trying to find a new app on disk and didn't have to foresight to save the output of info.

Reply:
> Also, will the info command recursively compile a list of ALL images/bundles referenced by a bundle, and, if so, do we have an idea for the scale we expect? Will it also be able to provide me info from an on disk bundle? If the scale is huge and I just spent a significant amount of time pulling everything, I probably wouldn't want to wait for imgpkg to essentially pull everything again just to create the list of images for the info command, especially if I need to do that every time I'm trying to find a new app on disk and didn't have the foresight to save the output of info.

Yes, the idea would be for the info command to provide information about all nested bundles.
I like the idea of the info command to read from disk, or even eventually we could think about the option of creating some sort of cache in disk with information on the images. But I will leave this idea for a later proposal.

> It just feels like a poor experience to have to find a specific digest in a list of potentially hundreds of images in order to find the app I'm looking for on disk.

Do you think that if the info command reads from disk this experience would be improved?


Action Item for me: Add the option for the info to read from disk instead.

Reply:
Just updated the proposal with 2 changes:
1. Add the command to read the information about a bundle that was previously pulled to disk
2. Add verbiage around the goals of the document and point out that in this proposal the changes are oriented to Applications used to interact with the bundle. Eventually, in a future proposal, we can create a better visualization of the bundles on disk to help with some of these UX problems.

Reply:
Reading an on disk bundle may not be useful since we prefer rewritting the imageslock if all the images are present in the bundle repo. This means users will most likely end up with the bundle repo name + digest, which isn't very informative since image names are lost there.

This may point back to the idea that had been proposed of somehow splitting the `.imgpkg` dir in to its own layer so that we can pull that easily without needing everything else in the image.

Reply:
> Reading an on disk bundle may not be useful since we prefer rewritting the imageslock if all the images are present in the bundle repo. This means users will most likely end up with the bundle repo name + digest, which isn't very informative since image names are lost there.

Yes, indeed you are right, that last change to the proposal might not be very useful after all. Going to revert that change.

Regarding the layering, I like the idea and we should create a proposal to define that.

Reply:
I believe this issue has been addressed in the newer review of the proposal